storage/imx6ull-flashnor/flashnor-test.c

```diff
@@ -0,0 +1,157 @@
+#include <getopt.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <meterfs.h>
+
+#include "storage.h"
+#include "flashnor-ecspi.h"
+
+
+static unsigned valcmp(uint8_t *buf, size_t len, uint8_t val)
+{
+	unsigned errors = 0;
+
+	for (size_t i = 0; i < len; i++) {
+		if (buf[i] != val)
+			errors++;
+	}
+
+	return errors;
+}
+
+
+void test_read_write(size_t flashsz, size_t sectorsz)
+{
+	int res = 0;
+	unsigned errors = 0;
+	uint8_t *pattern, *buf;
+
+	printf("%s: ", __func__);
+
+	pattern = malloc(sectorsz);
+	if (!pattern) {
+		printf("fail: malloc\n");
+		return;
+	}
+
+	buf = malloc(sectorsz);
+	if (!buf) {
+		printf("fail: malloc\n");
+		free(pattern);
+		return;
+	}
+
+	memset(pattern, 0x55, sectorsz);
+
+	/* Test only every sixteenth block */
+	for (unsigned int addr = 0; addr < flashsz; addr += sectorsz * 16) {
+		/* As meterfs is fixed right now in the driver implementation
+		   we have to erase sector before write, because of metadata written by meterfs */
+		if (flashnor_ecspiEraseSector(addr) < 0) {
+			printf("fail: erase sector");
+			res = -1;
+			break;
+		}
+
+		if (flashnor_ecspiWrite(addr, pattern, sectorsz) < 0) {
+			printf("fail: write sector at addr %u\n", addr);
+			res = -1;
+			break;
+		}
+
+		if (flashnor_ecspiRead(addr, buf, sectorsz) < 0) {
+			printf("fail: read sector at addr %u\n", addr);
+			res = -1;
+			break;
+		}
+
+		errors += valcmp(buf, sectorsz, pattern[0]);
+	}
+
+	if (res == 0) {
+		if (errors) {
+			printf("fail: read %u errors\n", errors);
```

Comment:
Is printf enough for this test? Maybe we would like to return error so we can mark the device as defective and stop any further testing?

Reply:
Now, I stop execution if test fails.